op-supervisor/supervisor/backend/rewinder/rewinder.go

```diff
@@ -0,0 +1,256 @@
+	for _, chainID := range r.db.DependencySet().Chains() {
+		if err := r.rewindL1ChainIfReorged(chainID, ev.IncomingBlock); err != nil {
+			r.log.Error("failed to rewind L1 data:", "chain", chainID, "err", err)
+		}
+	}
+}
+
+// handleLocalDerivedEvent checks if the newly derived block matches what we have in our unsafe DB
+// If it doesn't match, we need to rewind the logs DB to the common ancestor between
+// the LocalUnsafe head and the new LocalSafe block
+func (r *Rewinder) handleLocalDerivedEvent(ev superevents.LocalDerivedEvent) {
+	// Get the block at the derived height from our unsafe chain
+	newSafeHead := ev.Derived.Derived
+	unsafeVersion, err := r.db.FindSealedBlock(ev.ChainID, newSafeHead.Number)
+	if err != nil {
+		r.log.Error("failed to get unsafe block at derived height", "chain", ev.ChainID, "height", newSafeHead.Number, "err", err)
+		return
+	}
+
+	// If the block hashes match, our unsafe chain is still valid
+	if unsafeVersion.Hash == newSafeHead.Hash {
+		return
+	}
+
+	// Try rewinding the logs DB to the parent of the new safe head
+	// If it fails with a data conflict walk back through the chain
+	// until we find a common ancestor or reach the finalized block
+	finalized, err := r.db.Finalized(ev.ChainID)
+	if err != nil {
+		r.log.Error("failed to get finalized block", "chain", ev.ChainID, "err", err)
+		return
+	}
+	for height := int64(newSafeHead.Number - 1); height >= int64(finalized.Number); height-- {
+		// Get the block at this height
+		block, err := r.db.FindSealedBlock(ev.ChainID, uint64(height))
+		if err != nil {
+			r.log.Error("failed to get sealed block", "chain", ev.ChainID, "height", height, "err", err)
+			return
+		}
+
+		// Try to rewind and stop if it succeeds
+		err = r.db.RewindLogs(ev.ChainID, block)
+		if err == nil {
+			break
+		}
+
+		// If it failed with a data conflict try the next block
+		if errors.Is(err, types.ErrConflict) {
+			continue
+		}
+
+		// If it failed with any other error, log and return
+		r.log.Error("failed to rewind logs DB", "chain", ev.ChainID, "err", err)
+		return
+	}
+
+	// Emit event to trigger node reset with new heads
+	r.emitter.Emit(superevents.ChainRewoundEvent{ChainID: ev.ChainID})
+}
+
+// rewindL1ChainIfReorged rewinds the L1 chain for the given chain ID if a reorg is detected
+// It checks the local-safe head against the canonical L1 block at the same height
+func (r *Rewinder) rewindL1ChainIfReorged(chainID eth.ChainID, newTip eth.BlockID) error {
+	// Get the current CrossSafe head and its L1 block
+	crossSafe, err := r.db.CrossSafe(chainID)
+	if err != nil {
+		return fmt.Errorf("failed to get cross safe for chain %s: %w", chainID, err)
+	}
+	crossSafeL1, err := r.db.CrossDerivedFromBlockRef(chainID, crossSafe.Derived.ID())
+	if err != nil {
+		return fmt.Errorf("failed to get cross safe L1 block for chain %s: %w", chainID, err)
+	}
+
+	// If we're still on the canonical chain, nothing to do
+	if crossSafeL1.Hash == newTip.Hash {
+		return nil
+	}
```

Comment:
Since `newTip` comes from the L1 Accessor, it may point at an L1 head that simply hasn't been given to the node yet. Rather, I would think that in order to determine canonical chain, you'd want to:
- Get the latest derived pair from the Local Database
- Check the L1 for the block at that height (not the latest necessarily)
- If *those* don't match, then you know it changed.

EDIT: I see you're doing something like that below, I just don't fully understand the use of crossSafe as the heuristic I think.